### ISSUE TITLE
Segfaults / dangling pointers

### ISSUE THREAD

**Issue by sebpuetz** [huggingface/tokenizers#354]
~~~Python
import tokenizers
mod = tokenizers.models.BPE()
tok = tokenizers.Tokenizer(mod)
tok.encode("Test")
del tok
mod.encode(["Test"])
~~~

~~~
[1]    7487 segmentation fault (core dumped)  ipython
~~~

This happens because the `Tokenizer` steals the `Box` from the `Model` passed to the constructor and replaces it with a raw pointer to that `Box`. When the `Tokenizer` is dropped, the `Box` with the `Model` gets deallocated and the raw pointer in the container is dangling.

One workaround for this issue would be the use of a managed reference (e.g. `Arc` since threading is required) on the Rust-side. 

E.g. (untested):

~~~Rust

#[derive(Clone)]
pub struct PyModel {
    pub model: Arc<dyn tk::tokenizer::Model>,
}

impl Model for PyModel {
    [...]
}

#[pymethods]
impl Tokenizer {
    #[new]
    fn new(model: PyRefMut<PyModel>) -> PyResult<Self> {
        let tokenizer = tk::tokenizer::Tokenizer::new(model.clone());
        Ok(Tokenizer { tokenizer })
    }
}
~~~

------

Another issue with `Container` is that it dereference whatever pointer gets thrown in there, e.g. this will also cause segmentation faults:

~~~Rust
let c: Container<String>::Pointer(std::ptr::null_mut());
c.execute(|s| s.to_string());
~~~

**Comment by n1t0** [comment]
Yes you are totally right, the `Container` is not safe in many cases in the current state. Using an `Arc` seemed to complicate the API quite a lot last time I tried though. I don't really like the idea of having to use an `Arc` in the main library "just" because the bindings might need to handle different ownership rules, and I was hoping to find a better idea to deal with this. What do you think?

Concerning the second issue with `Container`, I don't think it should be a problem since this is something we only use internally, so as long as we don't write this kind of thing, we should be pretty safe.

Anyway, all of this has been bothering me since day one, but I didn't find a good solution yet. If you have any idea, I'd be happy to discuss it!

**Comment by n1t0** [comment]
Your proposal makes me think about something else though. Maybe we could keep an `Arc` of the `Tokenizer` on the Model (or any other part) when we attach it to the tokenizer. This way we could maybe ensure that it doesn't get dropped when Python gets rid of it.

**Comment by sebpuetz** [comment]
I had played around with the `Arc` solution some time last week but ended up in a similar situation that placing `Arc` in the main lib seemed strange.

It should be possible, however, to implement the `Model` trait for the Python wrapper type and put that into the `Tokenizer`. As far as I can tell, this wouldn't add additional indirection either: The `Container` solution already `Box`es a `Box`. I.e. `Tokenizer::model` would be realized as:
~~~Rust
Box(PyModel(Arc(dyn Model)))
~~~ 

The other option would be adding type parameters to the `Tokenizer` but considering the amount of trait objects in the `Tokenizer`, the signature would blow up and be inlegible.

> Your proposal makes me think about something else though. Maybe we could keep an `Arc` of the `Tokenizer` on the Model (or any other part) when we attach it to the tokenizer. This way we could maybe ensure that it doesn't get dropped when Python gets rid of it.

I think such circular referencing becomes tedious and adds too much complexity, I'd suggest giving implementing `Model` on the wrapper types a shot, I'll check if I have a somewhat usable branch.

> Concerning the second issue with `Container`, I don't think it should be a problem since this is something we only use internally, so as long as we don't write this kind of thing, we should be pretty safe.

It's a public API that can lead to UB, therefore it should either be marked `unsafe` or fixed. An easy protection against the null-pointer issue is adding a null-check. The inherent unsafety of `to_pointer` can't be fixed. There's **no** way to ensure that the `Box` outlives the raw pointer. Therefore, any API not controlling both the pointer and the `Box` cannot ensure soundness. 

I'm also not sure whether there is issues with threading with the `Container` implementation. As far as I can tell, any threads with the `Pointer` variant can `execute` the `Container` while at the same time the thread with the `Owned` variant can mutate the content of the container since it's an owned `Box`.

**Comment by sebpuetz** [comment]
I just pushed another commit to #355 that removes `Container` completely from the Python bindings by making the main library's `Tokenizer` generic over the components instead of `Box`ing trait objects. 

It's a huge amount of changed lines with tons of repetition / boilerplate. It's not meant to be reviewed this way, I can break it down into smaller, more manageable chunks if you are interested in that solution.

**Comment by sebpuetz** [comment]
I went ahead and split that huge commit into one commit per component. You can take a look at #355.

**Comment by sebpuetz** [comment]
Should be fixed by #355 #365!
